File: scripts/code-analyzer/parse-results.py
#!/usr/bin/env python3
"""
SMART Code Analyzer — Results Parser
=====================================
Reads the two CSV output files from run-analysis.sh and prints a
structured summary to the terminal and optionally to a text file.

Usage:
    python3 scripts/code-analyzer/parse-results.py \
        --pass1 reports/code-analysis.csv \
        --pass2 reports/code-analysis-with-sfge.csv \
        --out   reports/code-analysis-summary.txt
"""

import argparse
import collections
import csv
import sys
from datetime import datetime
from pathlib import Path


SEV_LABEL = {'1': 'Critical', '2': 'High', '3': 'Moderate', '4': 'Low', '5': 'Info'}


def load_csv(path: str) -> list[dict]:
    rows = []
    with open(path, newline='', encoding='utf-8') as f:
        for row in csv.DictReader(f):
            rows.append(row)
    return rows


def component_type(file_path: str) -> str:
    if '/classes/' in file_path:    return 'Apex Classes'
    if '/triggers/' in file_path:   return 'Triggers'
    if '/pages/' in file_path:      return 'Visualforce Pages'
    if '/aura/' in file_path:       return 'Aura'
    if '/lwc/' in file_path:        return 'LWC'
    if '/flows/' in file_path:      return 'Flows'
    return 'Other'


def bar(n: int, total: int, width: int = 30) -> str:
    filled = round(n / total * width) if total else 0
    return '█' * filled + '░' * (width - filled)


def section(title: str, lines: list[str]) -> list[str]:
    border = '─' * 60
    return ['', border, f'  {title}', border] + lines


def summarise(pass1_rows: list[dict], pass2_rows: list[dict]) -> list[str]:
    out: list[str] = []

    # ── Identify SFGE-only violations from Pass 2 ────────────────────────
    sfge_rows = [r for r in pass2_rows if r['engine'] == 'sfge']

    # Combined = Pass 1 (all) + SFGE delta
    combined_high     = sum(1 for r in pass1_rows if r['severity'] == '2') + \
                        sum(1 for r in sfge_rows  if r['severity'] == '2')
    combined_moderate = sum(1 for r in pass1_rows if r['severity'] == '3') + \
                        sum(1 for r in sfge_rows  if r['severity'] == '3')
    combined_total    = len(pass1_rows) + len(sfge_rows)

    # ── Header ────────────────────────────────────────────────────────────
    out += [
        '════════════════════════════════════════════════════════════',
        '  SMART — Salesforce Code Analyzer Summary',
        f'  Generated: {datetime.now().strftime("%Y-%m-%d %H:%M")}',
        '════════════════════════════════════════════════════════════',
        '',
        '  Two passes were run:',
        '  Pass 1 — Recommended rules, all component types',
        '  Pass 2 — Recommended + SFGE (data-flow), Apex/Triggers only',
    ]

    # ── Combined top-line ──────────────────────────────────────────────
    files1 = len({r['file'] for r in pass1_rows})
    files2 = len({r['file'] for r in sfge_rows})
    out += section('COMBINED TOTALS', [
        f'  Total violations : {combined_total:,}',
        f'  Files affected   : {files1 + files2:,}',
        f'  High severity    : {combined_high:,}',
        f'  Moderate severity: {combined_moderate:,}',
        f'  SFGE-only (new)  : {len(sfge_rows)}  ← not in Pass 1',
    ])

    # ── Pass 1 breakdown by severity ──────────────────────────────────
    by_sev = collections.Counter(SEV_LABEL.get(r['severity'], r['severity']) for r in pass1_rows)
    total1 = len(pass1_rows)
    sev_order = ['Critical', 'High', 'Moderate', 'Low', 'Info']
    sev_lines = [f'  {"Severity":<12} {"Count":>6}   {"Bar":<30}  {"Pct":>5}']
    sev_lines.append(f'  {"─"*12} {"─"*6}   {"─"*30}  {"─"*5}')
    for s in sev_order:
        n = by_sev.get(s, 0)
        if n:
            sev_lines.append(f'  {s:<12} {n:>6}   {bar(n, total1):<30}  {n/total1*100:>4.1f}%')
    out += section('PASS 1 — By Severity', sev_lines)

    # ── Pass 1 breakdown by engine ────────────────────────────────────
    by_eng = collections.Counter(r['engine'] for r in pass1_rows)
    eng_lines = [f'  {"Engine":<12} {"Total":>6}  {"High":>6}  {"Moderate":>8}  {"Scope"}']
    eng_lines.append(f'  {"─"*12} {"─"*6}  {"─"*6}  {"─"*8}  {"─"*20}')
    scope_map = {
        'pmd':      'Apex, VF Pages',
        'regex':    'All file types',
        'flow':     'Flows',
        'eslint':   'LWC, Aura JS',
        'cpd':      'Apex (copy-paste)',
        'retire-js':'JS (security)',
    }
    for eng, n in by_eng.most_common():
        high = sum(1 for r in pass1_rows if r['engine'] == eng and r['severity'] == '2')
        mod  = sum(1 for r in pass1_rows if r['engine'] == eng and r['severity'] == '3')
        scope = scope_map.get(eng, '')
        eng_lines.append(f'  {eng:<12} {n:>6}  {high:>6}  {mod:>8}  {scope}')
    out += section('PASS 1 — By Engine', eng_lines)

    # ── Pass 1 breakdown by component type ────────────────────────────
    by_comp = collections.Counter(component_type(r['file']) for r in pass1_rows)
    comp_lines = [f'  {"Component":<22} {"Count":>6}   {"Bar":<30}']
    comp_lines.append(f'  {"─"*22} {"─"*6}   {"─"*30}')
    for comp, n in by_comp.most_common():
        comp_lines.append(f'  {comp:<22} {n:>6}   {bar(n, total1):<30}')
    out += section('PASS 1 — By Component Type', comp_lines)

    # ── Pass 1 top rules ──────────────────────────────────────────────
    by_rule = collections.Counter(r['rule'] for r in pass1_rows)
    rule_lines = [f'  {"#":>3}  {"Rule":<50} {"Sev":>3}  {"Count":>6}']
    rule_lines.append(f'  {"─"*3}  {"─"*50} {"─"*3}  {"─"*6}')
    for i, (rule, n) in enumerate(by_rule.most_common(15), 1):
        sample_rows = [r for r in pass1_rows if r['rule'] == rule]
        sev = SEV_LABEL.get(sample_rows[0]['severity'], '?')[0] if sample_rows else '?'
        rule_lines.append(f'  {i:>3}  {rule:<50} {sev:>3}  {n:>6}')
    out += section('PASS 1 — Top 15 Rules by Count', rule_lines)

    # ── SFGE findings ─────────────────────────────────────────────────
    if sfge_rows:
        sfge_by_rule = collections.Counter(r['rule'] for r in sfge_rows)
        sfge_lines = [
            f'  Total SFGE violations: {len(sfge_rows)} '
            f'({sum(1 for r in sfge_rows if r["severity"]=="2")} High, '
            f'{sum(1 for r in sfge_rows if r["severity"]=="3")} Moderate)',
            '',
            f'  {"Rule":<50} {"Sev":>3}  {"Count":>6}',
            f'  {"─"*50} {"─"*3}  {"─"*6}',
        ]
        for rule, n in sfge_by_rule.most_common():
            sample = next(r for r in sfge_rows if r['rule'] == rule)
            sev = SEV_LABEL.get(sample['severity'], '?')
            sfge_lines.append(f'  {rule:<50} {sev[0]:>3}  {n:>6}')
        sfge_lines += ['', '  Top files with SFGE violations:']
        by_file = collections.Counter(r['file'].split('/')[-1] for r in sfge_rows)
        for fname, n in by_file.most_common(10):
            sfge_lines.append(f'    {fname:<45} {n:>3}')
        out += section('PASS 2 — SFGE (Salesforce Graph Engine) — NEW violations', sfge_lines)
    else:
        out += section('PASS 2 — SFGE', ['  No SFGE violations found, or SFGE CSV not provided.'])

    # ── Top files overall ──────────────────────────────────────────────
    all_rows = pass1_rows + sfge_rows
    by_file = collections.Counter(r['file'].split('/')[-1] for r in all_rows)
    file_lines = [f'  {"File":<50} {"Violations":>10}']
    file_lines.append(f'  {"─"*50} {"─"*10}')
    for fname, n in by_file.most_common(10):
        file_lines.append(f'  {fname:<50} {n:>10}')
    out += section('TOP 10 FILES BY VIOLATION COUNT (combined)', file_lines)

    # ── Quick-fix guide ───────────────────────────────────────────────
    out += section('RECOMMENDED REMEDIATION ORDER', [
        '',
        '  Priority 1 — Security (SFGE findings, fix immediately)',
        '    ApexFlsViolation          → add stripInaccessible() or WITH SECURITY_ENFORCED',
        '    DatabaseOperationsMustUseWithSharing → change class to "with sharing"',
        '    AvoidDatabaseOperationInLoop → move SOQL outside loops, use bulk queries',
        '',
        '  Priority 2 — Security (PMD/Flow)',
        '    PreventPassingUserDataIntoElementWithoutSharing → add sharing to Flows',
        '    ApexCRUDViolation         → add isAccessible()/isUpdateable() CRUD checks',
        '    AvoidHardcodingId         → use Custom Metadata or Custom Settings',
        '    AvoidOldSalesforceApiVersions → update metadata to API 63.0+',
        '',
        '  Priority 3 — Reliability',
        '    MissingNullCheckOnSoqlVariable → null-check after single-row SOQL',
        '    SameRecordUpdate (flows)  → consolidate updates in Flow elements',
        '    AvoidDebugStatements      → remove or guard System.debug() calls',
        '',
        '  Priority 4 — Quality & Formatting (autofix)',
        '    NoTrailingWhitespace      → run: npm run prettier',
        '    NoMixedIndentation        → run: npm run prettier',
        '    ApexUnitTestClassShouldHaveRunAs → wrap tests in System.runAs()',
        '    ApexDoc                   → add ApexDoc comments',
    ])

    out += ['', '═' * 60, '  End of report', '═' * 60, '']
    return out


def main():
    parser = argparse.ArgumentParser(description='Summarise Code Analyzer CSV outputs')
    parser.add_argument('--pass1', required=True, help='Path to Pass 1 CSV (Recommended, all components)')
    parser.add_argument('--pass2', required=True, help='Path to Pass 2 CSV (Recommended + SFGE, Apex only)')
    parser.add_argument('--out',   required=False, help='Optional output text file path')
    args = parser.parse_args()

    if not Path(args.pass1).exists():
        print(f'ERROR: Pass 1 CSV not found: {args.pass1}', file=sys.stderr)
        sys.exit(1)

    pass2_rows: list[dict] = []
    if Path(args.pass2).exists():
        pass2_rows = load_csv(args.pass2)
    else:
        print(f'WARNING: Pass 2 CSV not found: {args.pass2} — SFGE section will be empty.', file=sys.stderr)

    pass1_rows = load_csv(args.pass1)
    lines = summarise(pass1_rows, pass2_rows)
    output = '\n'.join(lines)

    print(output)

    if args.out:
        Path(args.out).write_text(output, encoding='utf-8')
        print(f'\nSummary written to: {args.out}')


if __name__ == '__main__':
    main()

File: scripts/code-analyzer/parse-results.js
#!/usr/bin/env node
/**
 * parse-results.js  —  Code Analyzer Results Parser
 *
 * Reads the two CSV output files from run-analysis.js (or run-analysis.sh)
 * and prints a structured violation summary to the terminal and optionally
 * to a text file.
 *
 * Usage:
 *   node scripts/code-analyzer/parse-results.js \
 *     --pass1 reports/code-analysis.csv \
 *     --pass2 reports/code-analysis-with-sfge.csv \
 *     --out   reports/code-analysis-summary.txt
 *
 * Defaults:
 *   --pass1  reports/code-analysis.csv
 *   --pass2  reports/code-analysis-with-sfge.csv
 *   --out    reports/code-analysis-summary.txt
 */

'use strict';

const fs   = require('fs');
const path = require('path');

// ─── Args ─────────────────────────────────────────────────────────────────────

const ROOT = process.cwd();

function arg(name, def) {
  const idx = process.argv.indexOf(name);
  return idx !== -1 && process.argv[idx + 1] ? process.argv[idx + 1] : def;
}

const PASS1_FILE = path.resolve(ROOT, arg('--pass1', 'reports/code-analysis.csv'));
const PASS2_FILE = path.resolve(ROOT, arg('--pass2', 'reports/code-analysis-with-sfge.csv'));
const OUT_FILE   = arg('--out', null) ? path.resolve(ROOT, arg('--out', null)) : null;

// ─── CSV Parser ───────────────────────────────────────────────────────────────

function parseCSV(filePath) {
  if (!fs.existsSync(filePath)) return null;

  const text  = fs.readFileSync(filePath, 'utf8');
  const lines = text.split(/\r?\n/);
  const headers = splitCSVLine(lines[0]);
  const rows = [];

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const values = splitCSVLine(line);
    const row = {};
    headers.forEach((h, j) => { row[h.trim()] = (values[j] || '').trim(); });
    rows.push(row);
  }
  return rows;
}

function splitCSVLine(line) {
  const result = [];
  let current  = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') { current += '"'; i++; }
      else { inQuotes = !inQuotes; }
    } else if (ch === ',' && !inQuotes) {
      result.push(current);
      current = '';
    } else {
      current += ch;
    }
  }
  result.push(current);
  return result;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

const SEV_LABEL = { '1': 'Critical', '2': 'High', '3': 'Moderate', '4': 'Low', '5': 'Info' };
const SEV_ORDER = ['Critical', 'High', 'Moderate', 'Low', 'Info'];

const SCOPE_MAP = {
  pmd:         'Apex, VF Pages',
  regex:       'All file types',
  flow:        'Flows',
  eslint:      'LWC, Aura JS',
  cpd:         'Apex (copy-paste)',
  'retire-js': 'JS (security)',
  sfge:        'Apex (path analysis)',
};

function componentType(filePath) {
  if (filePath.includes('/classes/'))  return 'Apex Classes';
  if (filePath.includes('/triggers/')) return 'Triggers';
  if (filePath.includes('/pages/'))    return 'Visualforce Pages';
  if (filePath.includes('/aura/'))     return 'Aura';
  if (filePath.includes('/lwc/'))      return 'LWC';
  if (filePath.includes('/flows/'))    return 'Flows';
  return 'Other';
}

function bar(n, total, width = 30) {
  const filled = total ? Math.round(n / total * width) : 0;
  return '█'.repeat(filled) + '░'.repeat(width - filled);
}

function section(title, lines) {
  const border = '─'.repeat(60);
  return ['', border, `  ${title}`, border, ...lines];
}

function counter(rows, key) {
  const map = {};
  rows.forEach((r) => { const v = r[key] || '?'; map[v] = (map[v] || 0) + 1; });
  return map;
}

function topN(map, n) {
  return Object.entries(map).sort((a, b) => b[1] - a[1]).slice(0, n);
}

// ─── Summary Builder ──────────────────────────────────────────────────────────

function summarise(pass1Rows, pass2Rows) {
  const out = [];

  const sfgeRows    = pass2Rows.filter((r) => r.engine === 'sfge');
  const combinedTotal = pass1Rows.length + sfgeRows.length;
  const combinedHigh  = pass1Rows.filter((r) => r.severity === '2').length +
                        sfgeRows.filter((r) => r.severity === '2').length;
  const combinedMod   = pass1Rows.filter((r) => r.severity === '3').length +
                        sfgeRows.filter((r) => r.severity === '3').length;
  const files1 = new Set(pass1Rows.map((r) => r.file)).size;
  const files2 = new Set(sfgeRows.map((r) => r.file)).size;

  const now = new Date().toISOString().replace('T', ' ').slice(0, 16);

  out.push(
    '════════════════════════════════════════════════════════════',
    '  Salesforce Code Analyzer Summary',
    `  Generated: ${now}`,
    '════════════════════════════════════════════════════════════',
    '',
    '  Two passes were run:',
    '  Pass 1 — Recommended rules, all component types',
    '  Pass 2 — Recommended + SFGE (data-flow), Apex/Triggers only',
  );

  // Combined totals
  out.push(...section('COMBINED TOTALS', [
    `  Total violations : ${combinedTotal.toLocaleString()}`,
    `  Files affected   : ${(files1 + files2).toLocaleString()}`,
    `  High severity    : ${combinedHigh.toLocaleString()}`,
    `  Moderate severity: ${combinedMod.toLocaleString()}`,
    `  SFGE-only (new)  : ${sfgeRows.length}  ← not in Pass 1`,
  ]));

  // Pass 1 by severity
  const bySev   = counter(pass1Rows, 'severity');
  const total1  = pass1Rows.length;
  const sevLines = [
    `  ${'Severity'.padEnd(12)} ${'Count'.padStart(6)}   ${'Bar'.padEnd(30)}  ${'Pct'.padStart(5)}`,
    `  ${'─'.repeat(12)} ${'─'.repeat(6)}   ${'─'.repeat(30)}  ${'─'.repeat(5)}`,
  ];
  SEV_ORDER.forEach((label) => {
    const key = Object.keys(SEV_LABEL).find((k) => SEV_LABEL[k] === label);
    const n   = bySev[key] || 0;
    if (n) sevLines.push(`  ${label.padEnd(12)} ${String(n).padStart(6)}   ${bar(n, total1).padEnd(30)}  ${(n / total1 * 100).toFixed(1).padStart(4)}%`);
  });
  out.push(...section('PASS 1 — By Severity', sevLines));

  // Pass 1 by engine
  const byEng    = counter(pass1Rows, 'engine');
  const engLines = [
    `  ${'Engine'.padEnd(12)} ${'Total'.padStart(6)}  ${'High'.padStart(6)}  ${'Moderate'.padStart(8)}  Scope`,
    `  ${'─'.repeat(12)} ${'─'.repeat(6)}  ${'─'.repeat(6)}  ${'─'.repeat(8)}  ${'─'.repeat(20)}`,
  ];
  topN(byEng, 10).forEach(([eng, n]) => {
    const high  = pass1Rows.filter((r) => r.engine === eng && r.severity === '2').length;
    const mod   = pass1Rows.filter((r) => r.engine === eng && r.severity === '3').length;
    const scope = SCOPE_MAP[eng] || '';
    engLines.push(`  ${eng.padEnd(12)} ${String(n).padStart(6)}  ${String(high).padStart(6)}  ${String(mod).padStart(8)}  ${scope}`);
  });
  out.push(...section('PASS 1 — By Engine', engLines));

  // Pass 1 by component type
  const byComp   = counter(pass1Rows.map((r) => ({ ...r, _comp: componentType(r.file || '') })), '_comp');
  const compLines = [
    `  ${'Component'.padEnd(22)} ${'Count'.padStart(6)}   Bar`,
    `  ${'─'.repeat(22)} ${'─'.repeat(6)}   ${'─'.repeat(30)}`,
  ];
  topN(byComp, 10).forEach(([comp, n]) => {
    compLines.push(`  ${comp.padEnd(22)} ${String(n).padStart(6)}   ${bar(n, total1)}`);
  });
  out.push(...section('PASS 1 — By Component Type', compLines));

  // Pass 1 top 15 rules
  const byRule   = counter(pass1Rows, 'rule');
  const ruleLines = [
    `  ${'#'.padStart(3)}  ${'Rule'.padEnd(50)} ${'Sev'.padStart(3)}  ${'Count'.padStart(6)}`,
    `  ${'─'.repeat(3)}  ${'─'.repeat(50)} ${'─'.repeat(3)}  ${'─'.repeat(6)}`,
  ];
  topN(byRule, 15).forEach(([rule, n], i) => {
    const sample  = pass1Rows.find((r) => r.rule === rule);
    const sevChar = sample ? (SEV_LABEL[sample.severity] || '?')[0] : '?';
    ruleLines.push(`  ${String(i + 1).padStart(3)}  ${rule.padEnd(50)} ${sevChar.padStart(3)}  ${String(n).padStart(6)}`);
  });
  out.push(...section('PASS 1 — Top 15 Rules by Count', ruleLines));

  // SFGE findings
  if (sfgeRows.length) {
    const sfgeByRule = counter(sfgeRows, 'rule');
    const sfgeHigh   = sfgeRows.filter((r) => r.severity === '2').length;
    const sfgeMod    = sfgeRows.filter((r) => r.severity === '3').length;
    const sfgeLines  = [
      `  Total SFGE violations: ${sfgeRows.length} (${sfgeHigh} High, ${sfgeMod} Moderate)`,
      '',
      `  ${'Rule'.padEnd(50)} ${'Sev'.padStart(3)}  ${'Count'.padStart(6)}`,
      `  ${'─'.repeat(50)} ${'─'.repeat(3)}  ${'─'.repeat(6)}`,
    ];
    topN(sfgeByRule, 20).forEach(([rule, n]) => {
      const sample  = sfgeRows.find((r) => r.rule === rule);
      const sevChar = sample ? (SEV_LABEL[sample.severity] || '?')[0] : '?';
      sfgeLines.push(`  ${rule.padEnd(50)} ${sevChar.padStart(3)}  ${String(n).padStart(6)}`);
    });
    sfgeLines.push('', '  Top files with SFGE violations:');
    const sfgeByFile = counter(sfgeRows.map((r) => ({ ...r, _f: (r.file || '').split('/').pop() })), '_f');
    topN(sfgeByFile, 10).forEach(([fname, n]) => {
      sfgeLines.push(`    ${fname.padEnd(45)} ${String(n).padStart(3)}`);
    });
    out.push(...section('PASS 2 — SFGE (Salesforce Graph Engine) — NEW violations', sfgeLines));
  } else {
    out.push(...section('PASS 2 — SFGE', ['  No SFGE violations found, or SFGE CSV not provided.']));
  }

  // Top 10 files overall
  const allRows   = [...pass1Rows, ...sfgeRows];
  const byFile    = counter(allRows.map((r) => ({ ...r, _f: (r.file || '').split('/').pop() })), '_f');
  const fileLines = [
    `  ${'File'.padEnd(50)} ${'Violations'.padStart(10)}`,
    `  ${'─'.repeat(50)} ${'─'.repeat(10)}`,
  ];
  topN(byFile, 10).forEach(([fname, n]) => {
    fileLines.push(`  ${fname.padEnd(50)} ${String(n).padStart(10)}`);
  });
  out.push(...section('TOP 10 FILES BY VIOLATION COUNT (combined)', fileLines));

  // Remediation guide
  out.push(...section('RECOMMENDED REMEDIATION ORDER', [
    '',
    '  Priority 1 — Security (SFGE findings, fix immediately)',
    '    ApexFlsViolation              → add stripInaccessible() or WITH SECURITY_ENFORCED',
    '    DatabaseOperationsMustUseWithSharing → change class to "with sharing"',
    '    AvoidDatabaseOperationInLoop  → move SOQL outside loops, use bulk queries',
    '',
    '  Priority 2 — Security (PMD/Flow)',
    '    PreventPassingUserDataIntoElementWithoutSharing → add sharing to Flows',
    '    ApexCRUDViolation             → add isAccessible()/isUpdateable() CRUD checks',
    '    AvoidHardcodingId             → use Custom Metadata or Custom Settings',
    '    AvoidOldSalesforceApiVersions → update metadata to API 63.0+',
    '',
    '  Priority 3 — Reliability',
    '    MissingNullCheckOnSoqlVariable → null-check after single-row SOQL',
    '    SameRecordUpdate (flows)      → consolidate updates in Flow elements',
    '    AvoidDebugStatements          → remove or guard System.debug() calls',
    '',
    '  Priority 4 — Quality & Formatting (autofix)',
    '    NoTrailingWhitespace          → run: npm run prettier',
    '    NoMixedIndentation            → run: npm run prettier',
    '    ApexUnitTestClassShouldHaveRunAs → wrap tests in System.runAs()',
  ]));

  out.push('', '═'.repeat(60), '  End of report', '═'.repeat(60), '');
  return out;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

function run() {
  if (!fs.existsSync(PASS1_FILE)) {
    console.error(`ERROR: Pass 1 CSV not found: ${PASS1_FILE}`);
    console.error('Run the code analyzer first to generate this file.');
    process.exit(1);
  }

  const pass1Rows = parseCSV(PASS1_FILE);
  const pass2Rows = fs.existsSync(PASS2_FILE) ? parseCSV(PASS2_FILE) : [];

  if (!fs.existsSync(PASS2_FILE)) {
    console.warn(`WARNING: Pass 2 CSV not found: ${PASS2_FILE} — SFGE section will be empty.`);
  }

  const lines  = summarise(pass1Rows, pass2Rows);
  const output = lines.join('\n');

  console.log(output);

  if (OUT_FILE) {
    fs.mkdirSync(path.dirname(OUT_FILE), { recursive: true });
    fs.writeFileSync(OUT_FILE, output, 'utf8');
    console.log(`\nSummary written to: ${OUT_FILE}`);
  }
}

run();
